torch/jit/_script.py

```diff
@@ -0,0 +1,169 @@
+        # Aggregate all the ScriptMethods and constants from superclasses
+        cls._methods = {}
+        cls._constants_set = set(getattr(cls, "__constants__", ()))
+        for base in reversed(bases):
+            for k, v in getattr(base, "_methods", {}).items():
+                cls._methods[k] = v
+            base_constants = getattr(base, "_constants_set", set())
+            cls._constants_set = cls._constants_set.union(base_constants)
+
+        # find all the script methods of the current class
+        for k, v in sorted(attrs.items()):
+            if isinstance(v, ScriptMethodStub):
+                delattr(cls, k)
+                cls._methods[v.original_method.__name__] = v
+
+        if getattr(cls, "_disable_script_meta", False):
+            # We leave built-in ScriptModule types alone, since this metaclass
+            # is only for compiling user classes that inherit from
+            # ScriptModule.
+            return super(ScriptMeta, cls).__init__(name, bases, attrs)
+
+        original_init = getattr(cls, "__init__", lambda self: None)
+
+        @functools.wraps(original_init)
+        def init_then_script(self, *args, **kwargs):
+            original_init(self, *args, **kwargs)
+            if type(self) == cls:
+
+                def make_stubs(module):
+                    cls = type(module)
+                    return [v for k, v in sorted(cls._methods.items())]
+
+                self.__dict__[
+                    "_actual_script_module"
+                ] = torch.jit._recursive.create_script_module(self, make_stubs)
+
+                # Delete the Python attributes that now shadow the ScriptModule
+                # ones, so that __getattr__ and __setattr__ will properly find
+                # the scripted versions.
+                concrete_type = self._actual_script_module._concrete_type
+                for name in concrete_type.get_attributes():
+                    delattr(self, name)
+                for name, _ in concrete_type.get_modules():
+                    delattr(self, name)
+                for name in ("_parameters", "_buffers", "_modules"):
+                    delattr(self, name)
+
+        cls.__init__ = init_then_script
+        return super(ScriptMeta, cls).__init__(name, bases, attrs)
+
+
+class _CachedForward(object):
+    def __get__(self, obj, cls):
+        return self.__getattr__("forward")
+
+
+if _enabled:
+    # this is a Python 'non-data descriptor' that causes the first access
+    # to ScriptModule's forward to lookup the forward method and stash
+    # it in the objects dict. Due to the standard rules for attribute lookup
+    # subsequent lookups will just directly return the previously looked up method.
+    # This is necessary because nn.Module defines forward as a method. If we
+    # did nothing __getattr__ would not be called. Instead we'd get nn.Module.forward
+    # which always throws an exception.
+
+    class ScriptModule(with_metaclass(ScriptMeta, Module)):
```

Comment:
should `RecursiveScriptModule` / `ScriptMeta` / `script` `script_method` be in this file ?

Reply:
see above—I can stack some more stuff on top so that we can get a better sense of what the end result looks like